src/java.net.http/share/classes/jdk/internal/net/http/hpack/QuickHuffman.java

```diff
@@ -735,7 +735,7 @@ public boolean write(ByteBuffer destination) {
                     long len = codeLengthOf(c);
                     if (bufferLen + len <= 64) {
                         buffer |= (codeValueOf(c) >>> bufferLen); // append
-                        bufferLen += len;
+                        bufferLen += (int) len;
                         pos++;
```

Comment:
Would it be cleaner to do the cast in the codeLengthOf method, so it returns an int?

Reply:
I believe the method returns an "unsigned int" - having the method return an int would then potentially cause `bufferLen + len <= 64` to yield true when it shouldn't. Hopefully @pavelrappo will comment.

Reply:
codeLengthOf() returns long. It could be changed to return int with a cast internally and then you could avoid the two new casts.

Reply:
No because the int returned could be negative, while the long will not. Assuming bufferLen is 0 and codeLengthOf() returns some value that has the 32th bit set to 1 then when codeLengthOf() returns long, bufferLen + codeLengthOf() will be a positive long > 64, and we won't enter the `if` here but if codeLengthOf() returns `int`, then bufferLen + codeLengthOf() would be negative and the `if` would be wrongly entered. I am not 100% sure this is a scenario that might occur (codeLengthOf() returning large "unsigned int" values) - but I'd prefer to stay on the safe side and assume that it can.

Reply:
This is what I mean:
```
jshell> long codeOfLength = (long)Integer.MAX_VALUE + 1
codeOfLength ==> 2147483648

jshell> int bufferLen = 0
bufferLen ==> 0

jshell> bufferLen + codeOfLength <= 64
$3 ==> false

jshell> bufferLen + (int)codeOfLength <= 64
$4 ==> true
```

Reply:
Yes, inserting explicit casts seems less clean than changing `codeLengthOf` to this:

    private static int codeLengthOf(char c) {
        return (int) (codes[c] & 0x00000000ffffffffL);
    }

There are 256 elements in constant `long[] codes`. One could easily check that each element when ANDed with `0x00000000ffffffffL` results in a value that fits into the first 31 bits of `int`.

Reply:
OK -  I will change codeLengthOf as suggested. It was not immediately obvious to me that the values would fit in the first 31 bits.

Reply:
> OK - I will change codeLengthOf as suggested. It was not immediately obvious to me that the values would fit in the first 31 bits.

In fact, it would even fit into the first 30 bits. There's a top-level comment that explains the layout of `code` elements. Maybe you can improve it somehow or carry over that specific part about the length into `codeLengthOf`. Alternatively, you can always slap `assert` in `codeLengthOf`.

Reply:
Did I understand correctly that the lower 32 bits is expected to contain a length coded on 5 bits - which is expected to be a value in (5..30)?

Reply:
Yes, you did understand that correctly.